File: packages/ios-sdk/Sources/PhosraSDK/Enforcement/EnforcementEngine.swift
import Foundation

#if canImport(FamilyControls)
import FamilyControls
#endif

/// Main orchestrator for applying a `CompiledPolicy` to the device using Apple's
/// FamilyControls, ManagedSettings, and DeviceActivity frameworks.
///
/// ## Usage
///
/// ```swift
/// let engine = EnforcementEngine()
///
/// // 1. Request FamilyControls authorization (once, during onboarding)
/// try await engine.requestAuthorization()
///
/// // 2. Apply a policy
/// let report = try engine.applyPolicy(compiledPolicy)
///
/// // 3. Check enforcement status
/// let status = engine.currentStatus()
/// ```
///
/// - Important: FamilyControls requires iOS 16+ and an Apple-provisioned entitlement.
///   The enforcement code below is scaffolding that compiles on macOS for development.
///   Real enforcement requires on-device testing with a signed build.
@available(iOS 16.0, macOS 13.0, *)
public final class EnforcementEngine {
    private let contentEnforcer: ContentFilterEnforcer
    private let screenTimeEnforcer: ScreenTimeEnforcer
    private let webFilterEnforcer: WebFilterEnforcer
    private let purchaseEnforcer: PurchaseEnforcer
    private let notificationEnforcer: NotificationEnforcer

    private var lastReport: EnforcementReport?

    public init() {
        self.contentEnforcer = ContentFilterEnforcer()
        self.screenTimeEnforcer = ScreenTimeEnforcer()
        self.webFilterEnforcer = WebFilterEnforcer()
        self.purchaseEnforcer = PurchaseEnforcer()
        self.notificationEnforcer = NotificationEnforcer()
    }

    /// Request FamilyControls authorization from the user.
    ///
    /// This must be called before any enforcement can take effect. On iOS, this
    /// presents a system dialog asking the parent/guardian to approve parental controls.
    ///
    /// - Throws: If the user denies authorization or FamilyControls is unavailable.
    public func requestAuthorization() async throws {
        #if canImport(FamilyControls) && os(iOS)
        try await AuthorizationCenter.shared.requestAuthorization(for: .individual)
        #else
        // MARK: - TODO: FamilyControls authorization is only available on iOS with the entitlement.
        // This no-op allows the SDK to compile on macOS for development.
        #endif
    }

    /// Apply a compiled policy to the device.
    ///
    /// Each policy section is applied independently. If one section fails, the others
    /// are still attempted. The returned `EnforcementReport` shows the status of each.
    ///
    /// - Parameter policy: The compiled policy from the Phosra API.
    /// - Returns: A report describing the enforcement outcome per category.
    public func applyPolicy(_ policy: CompiledPolicy) -> EnforcementReport {
        var report = EnforcementReport()

        report.contentFilter = contentEnforcer.apply(policy.contentFilter)
        report.screenTime = screenTimeEnforcer.apply(policy.screenTime)
        report.webFilter = webFilterEnforcer.apply(policy.webFilter)
        report.purchases = purchaseEnforcer.apply(policy.purchases)
        report.notifications = notificationEnforcer.apply(policy.notifications)

        self.lastReport = report
        return report
    }

    /// Remove all enforced restrictions from the device.
    ///
    /// This clears all ManagedSettingsStore settings, cancels DeviceActivity schedules,
    /// and removes notification curfew. Typically called when a device is unregistered.
    public func removeAllRestrictions() {
        contentEnforcer.remove()
        screenTimeEnforcer.remove()
        webFilterEnforcer.remove()
        purchaseEnforcer.remove()
        notificationEnforcer.remove()

        self.lastReport = nil
    }

    /// Get the most recent enforcement status.
    ///
    /// Returns the report from the last `applyPolicy` call, or a default pending report
    /// if no policy has been applied yet.
    public func currentStatus() -> EnforcementReport {
        return lastReport ?? EnforcementReport()
    }
}

// MARK: - Enforcement Report

/// Describes the outcome of applying a `CompiledPolicy` to the device.
public struct EnforcementReport: Sendable {
    public var contentFilter: CategoryReport = .init()
    public var screenTime: CategoryReport = .init()
    public var webFilter: CategoryReport = .init()
    public var purchases: CategoryReport = .init()
    public var notifications: CategoryReport = .init()

    public init() {}

    /// Converts the report into an array of `CategoryEnforcementResult` for API submission.
    public func toResults() -> [CategoryEnforcementResult] {
        return [
            CategoryEnforcementResult(
                category: "content_filter",
                status: contentFilter.status.rawValue,
                framework: contentFilter.framework,
                detail: contentFilter.details.isEmpty ? nil : contentFilter.details
            ),
            CategoryEnforcementResult(
                category: "screen_time",
                status: screenTime.status.rawValue,
                framework: screenTime.framework,
                detail: screenTime.details.isEmpty ? nil : screenTime.details
            ),
            CategoryEnforcementResult(
                category: "web_filter",
                status: webFilter.status.rawValue,
                framework: webFilter.framework,
                detail: webFilter.details.isEmpty ? nil : webFilter.details
            ),
            CategoryEnforcementResult(
                category: "purchases",
                status: purchases.status.rawValue,
                framework: purchases.framework,
                detail: purchases.details.isEmpty ? nil : purchases.details
            ),
            CategoryEnforcementResult(
                category: "notifications",
                status: notifications.status.rawValue,
                framework: notifications.framework,
                detail: notifications.details.isEmpty ? nil : notifications.details
            ),
        ]
    }

    /// Per-category enforcement result.
    public struct CategoryReport: Sendable {
        public var status: EnforcementStatus = .pending
        public var framework: String = ""
        public var details: String = ""
        public var errorMessage: String? = nil

        public init() {}
    }

    /// Enforcement status for a single category.
    public enum EnforcementStatus: String, Codable, Sendable {
        case pending
        case enforced
        case partial
        case failed
        case unsupported
    }
}

File: packages/android-sdk/src/main/kotlin/com/phosra/sdk/enforcement/EnforcementEngine.kt
package com.phosra.sdk.enforcement

import android.app.Activity
import android.app.AppOpsManager
import android.app.admin.DevicePolicyManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.VpnService
import android.os.Build
import android.os.Process
import android.provider.Settings
import com.phosra.sdk.models.CompiledPolicy
import com.phosra.sdk.services.PhosraDeviceAdminReceiver

/**
 * Main orchestrator for on-device policy enforcement.
 *
 * Coordinates all enforcement sub-systems:
 * - [AppBlockingEnforcer]: Blocks/allows apps via UsageStatsManager + overlay
 * - [ScreenTimeEnforcer]: Tracks and limits screen time via UsageStatsManager
 * - [WebFilterEnforcer]: DNS-based web content filtering via local VPN
 * - [PurchaseEnforcer]: Play Store purchase restrictions
 * - [NotificationEnforcer]: Notification curfew and filtering
 *
 * Usage:
 * ```kotlin
 * val engine = EnforcementEngine(context)
 * if (!engine.hasRequiredPermissions().allGranted) {
 *     engine.requestPermissions(activity)
 * }
 * val report = engine.applyPolicy(policy)
 * ```
 */
class EnforcementEngine(private val context: Context) {

    private val appBlocker = AppBlockingEnforcer(context)
    private val screenTime = ScreenTimeEnforcer(context)
    private val webFilter = WebFilterEnforcer(context)
    private val purchases = PurchaseEnforcer(context)
    private val notifications = NotificationEnforcer(context)

    /**
     * Apply a compiled policy to the device.
     * Each enforcement subsystem is applied independently; failures in one
     * do not prevent others from executing.
     *
     * @param policy The compiled policy from the Phosra API.
     * @return An [EnforcementReport] summarizing the result of each category.
     */
    fun applyPolicy(policy: CompiledPolicy): EnforcementReport {
        val contentFilterReport = try {
            appBlocker.apply(policy.contentFilter)
        } catch (e: Exception) {
            CategoryReport(
                status = EnforcementStatus.FAILED,
                framework = "UsageStatsManager",
                details = "Content filter enforcement failed",
                error = e
            )
        }

        val screenTimeReport = try {
            screenTime.apply(policy.screenTime)
        } catch (e: Exception) {
            CategoryReport(
                status = EnforcementStatus.FAILED,
                framework = "UsageStatsManager",
                details = "Screen time enforcement failed",
                error = e
            )
        }

        val webFilterReport = try {
            webFilter.apply(policy.webFilter)
        } catch (e: Exception) {
            CategoryReport(
                status = EnforcementStatus.FAILED,
                framework = "VpnService",
                details = "Web filter enforcement failed",
                error = e
            )
        }

        val purchasesReport = try {
            purchases.apply(policy.purchases)
        } catch (e: Exception) {
            CategoryReport(
                status = EnforcementStatus.FAILED,
                framework = "DevicePolicyManager",
                details = "Purchase enforcement failed",
                error = e
            )
        }

        val notificationsReport = try {
            notifications.apply(policy.notifications)
        } catch (e: Exception) {
            CategoryReport(
                status = EnforcementStatus.FAILED,
                framework = "NotificationListenerService",
                details = "Notification enforcement failed",
                error = e
            )
        }

        return EnforcementReport(
            contentFilter = contentFilterReport,
            screenTime = screenTimeReport,
            webFilter = webFilterReport,
            purchases = purchasesReport,
            notifications = notificationsReport
        )
    }

    /**
     * Remove all enforced restrictions and stop enforcement services.
     */
    fun removeAllRestrictions() {
        appBlocker.remove()
        screenTime.remove()
        webFilter.remove()
        purchases.remove()
        notifications.remove()
    }

    /**
     * Get the current enforcement status without applying any changes.
     */
    fun currentStatus(): EnforcementReport {
        return EnforcementReport(
            contentFilter = appBlocker.currentStatus(),
            screenTime = screenTime.currentStatus(),
            webFilter = webFilter.currentStatus(),
            purchases = purchases.currentStatus(),
            notifications = notifications.currentStatus()
        )
    }

    /**
     * Check which permissions have been granted for enforcement.
     */
    fun hasRequiredPermissions(): PermissionStatus {
        return PermissionStatus(
            usageStats = hasUsageStatsPermission(),
            overlay = hasOverlayPermission(),
            deviceAdmin = hasDeviceAdminPermission(),
            vpn = hasVpnPermission(),
            notifications = hasNotificationListenerPermission()
        )
    }

    /**
     * Launch system settings screens to request all required permissions.
     * Each permission requires its own system dialog / settings screen.
     *
     * @param activity The activity to launch permission requests from.
     */
    fun requestPermissions(activity: Activity) {
        val status = hasRequiredPermissions()

        if (!status.usageStats) {
            val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
            activity.startActivity(intent)
        }

        if (!status.overlay) {
            val intent = Intent(
                Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                android.net.Uri.parse("package:${context.packageName}")
            )
            activity.startActivity(intent)
        }

        if (!status.deviceAdmin) {
            val componentName = ComponentName(context, PhosraDeviceAdminReceiver::class.java)
            val intent = Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN).apply {
                putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN, componentName)
                putExtra(
                    DevicePolicyManager.EXTRA_ADD_EXPLANATION,
                    "Phosra needs device admin access to enforce parental controls."
                )
            }
            activity.startActivity(intent)
        }

        if (!status.vpn) {
            val intent = VpnService.prepare(activity)
            if (intent != null) {
                activity.startActivityForResult(intent, REQUEST_CODE_VPN)
            }
        }

        if (!status.notifications) {
            val intent = Intent(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS)
            activity.startActivity(intent)
        }
    }

    // ── Permission checks ──────────────────────────────────────────

    private fun hasUsageStatsPermission(): Boolean {
        val appOps = context.getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
        val mode = appOps.unsafeCheckOpNoThrow(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            Process.myUid(),
            context.packageName
        )
        return mode == AppOpsManager.MODE_ALLOWED
    }

    private fun hasOverlayPermission(): Boolean {
        return Settings.canDrawOverlays(context)
    }

    private fun hasDeviceAdminPermission(): Boolean {
        val dpm = context.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager
        val componentName = ComponentName(context, PhosraDeviceAdminReceiver::class.java)
        return dpm.isAdminActive(componentName)
    }

    private fun hasVpnPermission(): Boolean {
        // VPN permission is checked by calling VpnService.prepare()
        // which returns null if already granted. We can't check without an Activity,
        // so we assume granted if VPN was previously established.
        // TODO: Track VPN permission state in SharedPreferences after user grants it
        return true
    }

    private fun hasNotificationListenerPermission(): Boolean {
        val enabledListeners = Settings.Secure.getString(
            context.contentResolver,
            "enabled_notification_listeners"
        ) ?: return false
        return enabledListeners.contains(context.packageName)
    }

    companion object {
        const val REQUEST_CODE_VPN = 10001
    }
}

/**
 * Aggregated enforcement report across all categories.
 */
data class EnforcementReport(
    val contentFilter: CategoryReport = CategoryReport(),
    val screenTime: CategoryReport = CategoryReport(),
    val webFilter: CategoryReport = CategoryReport(),
    val purchases: CategoryReport = CategoryReport(),
    val notifications: CategoryReport = CategoryReport()
) {
    /**
     * Overall status: ENFORCED if all are enforced, PARTIAL if any are partial/failed,
     * FAILED if all have failed.
     */
    val overallStatus: EnforcementStatus
        get() {
            val statuses = listOf(
                contentFilter.status,
                screenTime.status,
                webFilter.status,
                purchases.status,
                notifications.status
            )
            return when {
                statuses.all { it == EnforcementStatus.ENFORCED } -> EnforcementStatus.ENFORCED
                statuses.all { it == EnforcementStatus.FAILED } -> EnforcementStatus.FAILED
                statuses.all { it == EnforcementStatus.PENDING } -> EnforcementStatus.PENDING
                else -> EnforcementStatus.PARTIAL
            }
        }
}

/**
 * Report for a single enforcement category.
 */
data class CategoryReport(
    val status: EnforcementStatus = EnforcementStatus.PENDING,
    val framework: String = "",
    val details: String = "",
    val error: Throwable? = null
)

/**
 * Status of a single enforcement category.
 */
enum class EnforcementStatus {
    /** Not yet attempted. */
    PENDING,

    /** Successfully enforced all rules in this category. */
    ENFORCED,

    /** Some rules enforced, some failed or unsupported. */
    PARTIAL,

    /** Enforcement failed entirely for this category. */
    FAILED,

    /** This enforcement category is not supported on this device/OS version. */
    UNSUPPORTED
}

/**
 * Status of all required permissions.
 */
data class PermissionStatus(
    val usageStats: Boolean = false,
    val overlay: Boolean = false,
    val deviceAdmin: Boolean = false,
    val vpn: Boolean = false,
    val notifications: Boolean = false
) {
    /** True if all required permissions are granted. */
    val allGranted: Boolean
        get() = usageStats && overlay && deviceAdmin && vpn && notifications

    /** List of permissions that are missing. */
    val missingPermissions: List<String>
        get() = buildList {
            if (!usageStats) add("Usage Stats Access")
            if (!overlay) add("Draw Over Other Apps")
            if (!deviceAdmin) add("Device Admin")
            if (!vpn) add("VPN")
            if (!notifications) add("Notification Listener")
        }
}
